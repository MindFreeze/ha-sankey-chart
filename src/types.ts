import {
  ActionConfig,
  BaseActionConfig,
  HapticType,
  LovelaceCard,
  LovelaceCardConfig,
  LovelaceCardEditor,
} from 'custom-card-helpers';
import { HassEntity, HassServiceTarget } from 'home-assistant-js-websocket';
import { UNIT_PREFIXES, CONVERSION_UNITS } from './const';

export interface SankeyChartConfig extends LovelaceCardConfig {
  type: string;
  autoconfig?: {
    print_yaml?: boolean;
  };
  title?: string;
  sections?: SectionConfig[];
  convert_units_to?: '' | CONVERSION_UNITS;
  co2_intensity_entity?: string;
  gas_co2_intensity?: number;
  monetary_unit?: string;
  electricity_price?: number;
  gas_price?: number;
  unit_prefix?: '' | keyof typeof UNIT_PREFIXES;
  round?: number;
  height?: number;
  wide?: boolean;
  layout?: 'auto' | 'vertical' | 'horizontal';
  show_icons?: boolean;
  show_names?: boolean;
  show_states?: boolean;
  show_units?: boolean;
  energy_date_selection?: boolean;
  min_box_size?: number;
  min_box_distance?: number;
  throttle?: number;
  min_state?: number;
  static_scale?: number;
  sort_by?: 'none' | 'state';
  sort_dir?: 'asc' | 'desc';
}

declare global {
  interface HTMLElementTagNameMap {
    'sankey-chart-editor': LovelaceCardEditor;
    'hui-error-card': LovelaceCard;
  }
}

export type BoxType = 'entity' | 'passthrough' | 'remaining_parent_state' | 'remaining_child_state';

export interface EntityConfig {
  entity_id: string;
  add_entities?: string[];
  subtract_entities?: string[];
  // @deprecated #100
  substract_entities?: string[];
  attribute?: string;
  type?: BoxType;
  children?: ChildConfigOrStr[];
  unit_of_measurement?: string; // for attribute
  color?: string;
  name?: string;
  icon?: string;
  color_on_state?: boolean;
  color_above?: string;
  color_below?: string;
  color_limit?: number;
  tap_action?: ActionConfigExtended;
  double_tap_action?: ActionConfigExtended;
  hold_action?: ActionConfigExtended;
  // @deprecated
  remaining?:
    | string
    | {
        name: string;
        color?: string;
      };
}

export type EntityConfigInternal = EntityConfig & {
  children: ChildConfigOrStr[];
  accountedState?: number;
  foundChildren?: string[];
};

export type EntityConfigOrStr = string | EntityConfig;

export type ChildConfig = {
  entity_id: string;
  connection_entity_id: string;
};

export type ChildConfigOrStr = string | ChildConfig;

export type ActionConfigExtended = ActionConfig | CallServiceActionConfig | MoreInfoActionConfig | ZoomActionConfig;

export interface MoreInfoActionConfig extends BaseActionConfig {
  action: 'more-info';
  entity?: string;
  data?: {
    entity_id?: string | [string];
  };
}

export interface ZoomActionConfig extends BaseActionConfig {
  action: 'zoom';
}

export interface CallServiceActionConfig extends BaseActionConfig {
  action: 'call-service';
  service: string;
  data?: {
    entity_id?: string | [string];
    [key: string]: unknown;
  };
  target?: HassServiceTarget;
  repeat?: number;
  haptic?: HapticType;
}

export interface SectionConfig {
  entities: EntityConfigOrStr[];
  sort_by?: 'none' | 'state';
  sort_dir?: 'asc' | 'desc';
  sort_group_by_parent?: boolean;
  min_width?: number;
}

export interface Section {
  entities: EntityConfigInternal[];
  sort_by?: 'none' | 'state';
  sort_dir?: 'asc' | 'desc';
  sort_group_by_parent?: boolean;
  min_width?: number;
}

export interface Config extends SankeyChartConfig {
  layout: 'auto' | 'vertical' | 'horizontal';
  unit_prefix: '' | keyof typeof UNIT_PREFIXES;
  round: number;
  height: number;
  min_box_size: number;
  min_box_distance: number;
  min_state: number;
  sections: Section[];
}

export interface Connection {
  startY: number;
  startSize: number;
  endY: number;
  endSize: number;
  state: number;
  startColor?: string;
  endColor?: string;
  highlighted?: boolean;
}

export interface Box {
  config: EntityConfigInternal;
  entity: Omit<HassEntity, 'state'> & {
    state: string | number;
  };
  entity_id: string;
  state: number;
  unit_of_measurement?: string;
  children: ChildConfigOrStr[];
  color: string;
  size: number;
  top: number;
  extraSpacers?: number;
  connections: {
    parents: Connection[];
  };
  connectedParentState: number;
}

export interface SectionState {
  boxes: Box[];
  total: number;
  spacerSize: number;
  statePerPixel: number;
  config: Section;
  size: number;
}

export interface ConnectionState {
  parent: EntityConfigInternal;
  child: EntityConfigInternal;
  state: number;
  prevParentState: number;
  prevChildState: number;
  ready: boolean;
  calculating?: boolean;
  highlighted?: boolean;
  passthroughs: EntityConfigInternal[];
}

export interface NormalizedState {
  state: number;
  unit_of_measurement?: string;
}
