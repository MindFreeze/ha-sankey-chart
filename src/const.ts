import { EntityConfig } from "./types";

export const UNIT_PREFIXES = {
  'm': 0.001,
  'k': 1000,
  'M': 1000000,
  'G': 1000000000,
  'T': 1000000000000,
};

export const MIN_LABEL_HEIGHT = 15;

export const DEFAULT_ENTITY_CONF: Omit<EntityConfig, 'entity_id'> = {
  type: 'entity',
};

export const FT3_PER_M3 = 35.31;
