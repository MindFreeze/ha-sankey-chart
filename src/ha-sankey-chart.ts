/* eslint-disable @typescript-eslint/no-explicit-any */
import { LitElement, html, TemplateResult } from 'lit';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { customElement, property, query, state } from 'lit/decorators';

import type { Config, EntityConfigInternal, SankeyChartConfig, Section } from './types';
import { version } from '../package.json';
import { localize } from './localize/localize';
import { normalizeConfig, renderError } from './utils';
import { SubscribeMixin } from './subscribe-mixin';
import './chart';
import { HassEntities } from 'home-assistant-js-websocket';
import {
  Conversions,
  EnergyCollection,
  ENERGY_SOURCE_TYPES,
  getEnergyDataCollection,
  getEnergySourceColor,
  getStatistics,
} from './energy';
import { until } from 'lit/directives/until';
import { getEntitiesByArea, HomeAssistantReal } from './hass';
import { LovelaceCardEditor } from 'custom-card-helpers';
import './editor/index';

/* eslint no-console: 0 */
console.info(
  `%c sankey-chart %c ${localize('common.version')} ${version} `,
  'color: orange; font-weight: bold; background: black',
  'color: white; font-weight: bold; background: dimgray',
);

// This puts your card into the UI card picker dialog
(window as any).customCards = (window as any).customCards || [];
(window as any).customCards.push({
  type: 'sankey-chart',
  name: 'Sankey Chart',
  description:
    'A card to display a sankey chart. For example for power consumptionA template custom card for you to create something awesome',
});

const ENERGY_DATA_TIMEOUT = 10000;

@customElement('sankey-chart')
class SankeyChart extends SubscribeMixin(LitElement) {
  public static async getConfigElement(): Promise<LovelaceCardEditor> {
    // await import('./editor');
    return document.createElement('sankey-chart-editor');
  }

  public static getStubConfig(): Record<string, unknown> {
    return { autoconfig: { print_yaml: false } };
  }

  // https://lit.dev/docs/components/properties/
  @property({ attribute: false }) public hass!: HomeAssistantReal;

  @state() private config!: Config;
  @state() private states: HassEntities = {};
  @state() private entityIds: string[] = [];
  @state() private error?: Error | unknown;
  @state() private forceUpdateTs?: number;

  public hassSubscribe() {
    if (!this.config.energy_date_selection) {
      return [];
    }
    const start = Date.now();
    const getEnergyDataCollectionPoll = (
      resolve: (value: EnergyCollection | PromiseLike<EnergyCollection>) => void,
      reject: (reason?: any) => void,
    ) => {
      const energyCollection = getEnergyDataCollection(this.hass);
      if (energyCollection) {
        resolve(energyCollection);
      } else if (Date.now() - start > ENERGY_DATA_TIMEOUT) {
        console.debug(getEnergyDataCollection(this.hass));
        reject(
          new Error('No energy data received. Make sure to add a `type: energy-date-selection` card to this screen.'),
        );
      } else {
        setTimeout(() => getEnergyDataCollectionPoll(resolve, reject), 100);
      }
    };
    const energyPromise = new Promise<EnergyCollection>(getEnergyDataCollectionPoll);
    setTimeout(() => {
      if (!this.error && !Object.keys(this.states).length) {
        this.error = new Error('Something went wrong. No energy data received.');
        console.debug(getEnergyDataCollection(this.hass));
      }
    }, ENERGY_DATA_TIMEOUT * 2);
    energyPromise.catch(err => {
      this.error = err;
    });
    return [
      energyPromise.then(async collection => {
        const isAutoconfig = this.config.autoconfig || typeof this.config.autoconfig === 'object';
        if (isAutoconfig && !this.config.sections.length) {
          try {
            await this.autoconfig(collection);
          } catch (err: any) {
            this.error = new Error(err?.message || err);
          }
        }
        return collection.subscribe(async data => {
          if (isAutoconfig && !this.config.sections.length) {
            try {
              await this.autoconfig(collection);
            } catch (err: any) {
              this.error = new Error(err?.message || err);
              return;
            }
          }
          if (this.entityIds.length) {
            const conversions: Conversions = {
              convert_units_to: this.config.convert_units_to!,
              co2_intensity_entity: this.config.co2_intensity_entity!,
              gas_co2_intensity: this.config.gas_co2_intensity!,
              electricity_price: this.config.electricity_price,
              gas_price: this.config.gas_price,
            };
            const stats = await getStatistics(this.hass, data, this.entityIds, conversions);
            const states: HassEntities = {};
            Object.keys(stats).forEach(id => {
              if (this.hass.states[id]) {
                states[id] = { ...this.hass.states[id], state: String(stats[id]) };
              }
            });
            this.states = states;
          }
          this.forceUpdateTs = Date.now();
        });
      }),
    ];
  }

  // https://lit.dev/docs/components/properties/#accessors-custom
  public setConfig(config: SankeyChartConfig, isMetric: boolean): void {
    if (typeof config !== 'object') {
      throw new Error(localize('common.invalid_configuration'));
    }

    this.setNormalizedConfig(normalizeConfig(config, isMetric));
    this.resetSubscriptions();
  }

  private setNormalizedConfig(config: Config): void {
    this.config = config;

    this.entityIds = [];
    this.config.sections.forEach(({ entities }) => {
      entities.forEach(ent => {
        if (ent.type === 'entity') {
          this.entityIds.push(ent.entity_id);
        }
        if (ent.add_entities) {
          ent.add_entities.forEach(e => this.entityIds.push(e));
        }
        if (ent.subtract_entities) {
          ent.subtract_entities.forEach(e => this.entityIds.push(e));
        }
      });
    });
  }

  private async autoconfig(collection: EnergyCollection) {
    if (!collection.prefs) {
      return;
    }
    const sources = (collection.prefs?.energy_sources || [])
      .map(s => ({
        ...s,
        ids: [s, ...(s.flow_from || [])]
          .map(f => f.stat_energy_from)
          .filter(id => {
            if (!id || !this.hass.states[id]) {
              if (id) {
                console.warn('Ignoring missing entity ' + id);
              }
              return false;
            }
            return true;
          }) as string[],
      }))
      .filter(s => ENERGY_SOURCE_TYPES.includes(s.type) && s.ids.length)
      .sort((s1, s2) => {
        // sort to solar, battery, grid
        if (s1.type === s2.type) {
          return 0;
        }
        if (s1.type === 'solar') {
          return -1;
        }
        if (s1.type === 'battery' && s2.type !== 'solar') {
          return -1;
        }
        return 1;
      });
    const deviceIds = (collection.prefs?.device_consumption || [])
      .map(d => d.stat_consumption)
      .filter(id => {
        if (!this.hass.states[id]) {
          console.warn('Ignoring missing entity ' + id);
          return false;
        }
        return true;
      });
    const areasResult = await getEntitiesByArea(this.hass, deviceIds);
    const areas = Object.values(areasResult)
      // put 'No area' last
      .sort((a, b) => (a.area.name === 'No area' ? 1 : b.area.name === 'No area' ? -1 : 0));
    const orderedDeviceIds = areas.reduce((all: string[], a) => [...all, ...a.entities], []);

    const sections = [
      {
        entities: sources.map(source => {
          const subtract = source.stat_energy_to
            ? [source.stat_energy_to]
            : source.flow_to?.map(e => e.stat_energy_to) || undefined;
          return {
            entity_id: source.ids[0],
            add_entities: source.ids?.length > 1 ? source.ids.slice(1) : undefined,
            subtract_entities: subtract,
            type: 'entity',
            color: getEnergySourceColor(source.type),
            children: ['total'],
          };
        }),
      },
      {
        entities: [
          {
            entity_id: 'total',
            type: sources.length ? 'remaining_parent_state' : 'remaining_child_state',
            name: 'Total Consumption',
            children: [...areas.map(a => a.area.area_id), 'unknown'],
          },
        ],
      },
      {
        entities: [
          ...areas.map(
            ({ area, entities }): EntityConfigInternal => ({
              entity_id: area.area_id,
              type: 'remaining_child_state',
              name: area.name,
              children: entities,
            }),
          ),
          {
            entity_id: 'unknown',
            type: 'remaining_parent_state',
            name: 'Unknown',
            children: [],
          },
        ],
      },
      {
        entities: orderedDeviceIds.map(id => ({
          entity_id: id,
          type: 'entity',
          children: [],
        })),
      },
    ].filter(s => s.entities.length > 0) as Section[];

    const grid = sources.find(s => s.type === 'grid');
    if (grid && grid?.flow_to?.length) {
      // grid export
      grid?.flow_to.forEach(({ stat_energy_to }) => {
        sections[1].entities.unshift({
          entity_id: stat_energy_to,
          subtract_entities: (grid.flow_from || []).map(e => e.stat_energy_from),
          type: 'entity',
          color: getEnergySourceColor(grid.type),
          children: [],
        });
        sections[0].entities.forEach(entity => {
          entity.children.unshift(stat_energy_to);
        });
      });
    }

    const battery = sources.find(s => s.type === 'battery');
    if (battery && battery.stat_energy_from && battery.stat_energy_to) {
      // battery charging
      sections[1].entities.unshift({
        entity_id: battery.stat_energy_to,
        subtract_entities: [battery.stat_energy_from],
        type: 'entity',
        color: getEnergySourceColor(battery.type),
        children: [],
      });
      sections[0].entities.forEach(entity => {
        entity.children.unshift(battery.stat_energy_to!);
      });
    }

    this.setNormalizedConfig({ ...this.config, sections });
  }

  // https://lit.dev/docs/components/rendering/
  protected render(): TemplateResult | void {
    if (this.error) {
      console.error(this.error);
      return html`${until(renderError(String(this.error), this.config, this.hass))}`;
    }
    const print_yaml = this.config.autoconfig?.print_yaml;
    return html`
      <sankey-chart-base
        .hass=${this.hass}
        .states=${this.config.energy_date_selection ? this.states : this.hass.states}
        .config=${this.config}
        .forceUpdateTs=${this.forceUpdateTs}
        .width=${this.clientWidth || this.offsetWidth || this.parentElement?.clientWidth || window.innerWidth}
      ></sankey-chart-base>
      ${print_yaml && this.config.sections.length
        ? html`${until(renderError('', { ...this.config, autoconfig: undefined }, this.hass))}`
        : ''}
    `;
  }
}

export default SankeyChart;