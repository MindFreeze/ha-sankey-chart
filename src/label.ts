import { html } from 'lit';
import { Box, Config } from './types';
import { styleMap } from 'lit/directives/style-map';
import { CHAR_WIDTH_RATIO, MIN_LABEL_HEIGHT } from './const';

export function renderLabel(
  box: Box,
  config: Config,
  formattedState: string,
  name: string,
  spacerSize: number,
  vertical: boolean,
) {
  const { show_names, show_states, show_units } = config;
  const shouldShowLabel = box.config.type !== 'passthrough' && (show_names || show_states);
  if (!shouldShowLabel) return null;

  const maxLabelSize = box.size + spacerSize - 1;

  // reduce label size if it doesn't fit
  const labelStyle: Record<string, string> = { lineHeight: MIN_LABEL_HEIGHT + 'px' };
  const nameStyle: Record<string, string> = {};
  if (vertical) {
    // count chars in the name and reduce font size if it doesn't fit maxLabelSize
    labelStyle.width = maxLabelSize + 'px';
    const stateChars = (formattedState + (show_units ? box.unit_of_measurement : '')).length;
    const desiredWidth = stateChars * CHAR_WIDTH_RATIO;
    if (desiredWidth > maxLabelSize) {
      const fontSize = maxLabelSize / desiredWidth;
      labelStyle.fontSize = `${fontSize}em`;
      labelStyle.lineHeight = `${fontSize}em`;
    }
    if (show_names) {
      const nameChars = Math.max(...name.split(/[\s]+/).map(l => l.length));
      const desiredNameWidth = nameChars * CHAR_WIDTH_RATIO;
      if (desiredNameWidth > maxLabelSize) {
        const fontSize = maxLabelSize / desiredNameWidth;
        nameStyle.fontSize = `${fontSize}rem`;
        nameStyle.lineHeight = `${fontSize}rem`;
      }
    }
  } else {
    if (maxLabelSize < MIN_LABEL_HEIGHT) {
      const fontSize = maxLabelSize / MIN_LABEL_HEIGHT;
      // labelStyle.maxHeight = maxLabelSize + 'px';
      labelStyle.fontSize = `${fontSize}em`;
      labelStyle.lineHeight = `${fontSize}em`;
    }
    const numLines = name.split('\n').filter(v => v).length;
    if (numLines > 1) {
      nameStyle.whiteSpace = 'pre';
      if (labelStyle.fontSize) {
        nameStyle.fontSize = `${1 / numLines + 0.1}rem`;
        nameStyle.lineHeight = `${1 / numLines + 0.1}rem`;
      } else if (maxLabelSize < MIN_LABEL_HEIGHT * numLines) {
        nameStyle.fontSize = `${(maxLabelSize / MIN_LABEL_HEIGHT / numLines) * 1.1}em`;
        nameStyle.lineHeight = `${(maxLabelSize / MIN_LABEL_HEIGHT / numLines) * 1.1}em`;
      }
    }
  }
  return html`<div class="label" style=${styleMap(labelStyle)}>
    ${show_states
      ? html`<span>
          <span class="state">${formattedState}</span>${show_units
            ? html`<span class="unit">${box.unit_of_measurement}</span>`
            : null}
        </span>`
      : null}
    ${show_names
      ? html`${!vertical ? html`&nbsp;` : null}<span class="name" style=${styleMap(nameStyle)}>${name}</span>`
      : null}
  </div>`;
}
