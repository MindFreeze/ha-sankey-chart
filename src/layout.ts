import { MIN_HORIZONTAL_SECTION_W } from './const';
import { Config } from './types';

export function shouldBeVertical(config: Config, width: number) {
  if (config.layout === 'auto') {
    const minWidth = config.sections.reduce((acc, section) => {
      const width = section.min_width ?? MIN_HORIZONTAL_SECTION_W;
      return width + acc;
    }, 0);
    return width < minWidth;
  }
  return config.layout === 'vertical';
}
