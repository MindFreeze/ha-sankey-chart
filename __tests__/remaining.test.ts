// import '../dist/ha-sankey-chart';
import { HomeAssistant } from 'custom-card-helpers';
import '../src/ha-sankey-chart';
import SankeyChart from '../src/ha-sankey-chart';
import { SankeyChartConfig } from '../src/types';
import mockHass from './__mocks__/hass.mock';
import { LitElement } from 'lit';

const hass = mockHass();

describe('SankeyChart with remaining type entities', () => {
  const ROOT_TAG = 'sankey-chart';
  let sankeyChart: SankeyChart;

  beforeEach(() => {
    sankeyChart = window.document.createElement(ROOT_TAG) as SankeyChart;
    // @ts-ignore
    sankeyChart.hass = hass as HomeAssistant;
  });

  afterEach(() => {
    sankeyChart.remove();
  });

  it('matches snapshot', async () => {
    const config: SankeyChartConfig = {
      type: 'custom:sankey-chart',
      title: 'Remaining test',
      show_states: true,
      min_state: 0.1,
      unit_prefix: 'k',
      sections: [
        {
          entities: [
            {
              entity_id: 'sensor.test_power',
              name: 'Total',
              color: 'var(--warning-color)',
              children: ['tt', 'sensor.test_power3', 'Annet'],
            },
          ],
        },
        {
          entities: [
            {
              entity_id: 'tt',
              type: 'remaining_child_state',
              name: 'Total\nAll\nAll\nAll\nAll\nAll\nIDK\nIDK\nIDK\nIDK\nIDK\nIDK\nIDK\nIDK\nIDK',
              color: 'var(--warning-color)',
              children: ['sensor.test_power1', 'sensor.test_power2', 'sensor.test_power4'],
              color_on_state: true,
              color_limit: 10.1,
              color_below: 'darkslateblue',
            },
          ],
        },
        {
          entities: [
            {
              entity_id: 'sensor.test_power1',
              name: 'Varmtvann\nBlaa',
              children: ['sensor.test_power3'],
            },
            {
              entity_id: 'sensor.test_power2',
              name: 'Avfukter',
              unit_of_measurement: 'Ð’',
              children: ['sensor.test_power3'],
            },
            {
              entity_id: 'sensor.test_power4',
              children: ['sensor.test_power3'],
            },
            {
              entity_id: 'Annet',
              type: 'remaining_child_state',
              name: 'Annet',
              children: ['sensor.test_power3'],
            },
          ],
        },
        {
          entities: [
            {
              entity_id: 'switch.plug_158d00022adfd9',
              attribute: 'load_power',
              unit_of_measurement: 'Wh',
              tap_action: {
                action: 'toggle',
              },
            },
          ],
        },
        {
          entities: [
            {
              entity_id: 'sensor.test_power3',
              color_below: 'red',
              color: 'red',
              color_limit: 14000,
            },
          ],
        },
      ],
    };
    sankeyChart.setConfig(config, true);
    document.body.appendChild(sankeyChart);
    await sankeyChart.updateComplete;

    expect(sankeyChart.shadowRoot?.innerHTML.replace(/<!--.*-->/g, '')).toMatchSnapshot();

    // Wait for the sankey-chart-base component to finish updating
    const sankeyChartBase = sankeyChart.shadowRoot?.querySelector('sankey-chart-base') as LitElement;
    expect(sankeyChartBase).not.toBeNull();

    expect(sankeyChartBase.shadowRoot?.innerHTML.replace(/<!--.*-->/g, '')).toMatchSnapshot();
  });
});
