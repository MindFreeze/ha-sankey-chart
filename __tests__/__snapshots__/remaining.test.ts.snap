// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SankeyChart with remaining type entities matches snapshot 1`] = `
"
      <sankey-chart-base></sankey-chart-base>
      
    "
`;

exports[`SankeyChart with remaining type entities matches snapshot 2`] = `
"
        <ha-card label="Sankey Chart">
          <div class=" container with-header " style="height:200px;">
            
    <div class="section" style="">
      <div class="connectors">
            <svg preserveAspectRatio="none" viewBox="0 0 100 200">
              
        <defs>
          
            <linearGradient id="gradient0.0.0">
              <stop offset="0%" stop-color="var(--warning-color)"></stop>
              <stop offset="100%" stop-color="darkslateblue"></stop>
            </linearGradient>
          
        </defs>
        
          <path d="M0,0 C50,0 50,0 100,0 L100,61.62162162162163 C50,61.62162162162163 50,61 0,61 Z" fill="url(#gradient0.0.0)" fill-opacity="0.4"></path>
        
      
            </svg>
          </div>
      
          
          
          <div class="box type-entity" style="height:61px;">
            <div style="background-color:var(--warning-color);" title="2kW Total" class="">
              
            </div>
            <div class="label" style="line-height:15px;">
                  
                  
                </div>
          </div>
          
        
    </div>
  
    <div class="section" style="">
      <div class="connectors">
            <svg preserveAspectRatio="none" viewBox="0 0 100 200">
              
        <defs>
          
            <linearGradient id="gradient1.0.0">
              <stop offset="0%" stop-color="darkslateblue"></stop>
              <stop offset="100%" stop-color="var(--primary-color)"></stop>
            </linearGradient>
          
            <linearGradient id="gradient1.0.1">
              <stop offset="0%" stop-color="darkslateblue"></stop>
              <stop offset="100%" stop-color="var(--primary-color)"></stop>
            </linearGradient>
          
            <linearGradient id="gradient1.0.2">
              <stop offset="0%" stop-color="darkslateblue"></stop>
              <stop offset="100%" stop-color="var(--primary-color)"></stop>
            </linearGradient>
          
        </defs>
        
          <path d="M0,0 C50,0 50,0 100,0 L100,63 C50,63 50,63.2879899767317 0,63.2879899767317 Z" fill="url(#gradient1.0.0)" fill-opacity="0.4"></path>
        
          <path d="M0,63.2879899767317 C50,63.2879899767317 50,68.5 100,68.5 L100,131.5 C50,131.5 50,126.60998747091463 0,126.60998747091463 Z" fill="url(#gradient1.0.1)" fill-opacity="0.4"></path>
        
          <path d="M0,126.60998747091462 C50,126.60998747091462 50,137 100,137 L100,200 C50,200 50,190 0,190 Z" fill="url(#gradient1.0.2)" fill-opacity="0.4"></path>
        
      
            </svg>
          </div>
      
          
          
          <div class="box type-remaining_child_state" style="height:190px;">
            <div style="background-color:darkslateblue;" title="6kW Total
All
All
All
All
All
IDK
IDK
IDK
IDK
IDK
IDK
IDK
IDK
IDK" class="">
              
            </div>
            <div class="label" style="line-height:15px;">
                  
                  
                </div>
          </div>
          
        
    </div>
  
    <div class="section" style="">
      <div class="connectors">
            <svg preserveAspectRatio="none" viewBox="0 0 100 200">
              
        <defs>
          
            <linearGradient id="gradient2.0.0">
              <stop offset="0%" stop-color="var(--primary-color)"></stop>
              <stop offset="100%" stop-color="red"></stop>
            </linearGradient>
          
        </defs>
        
          <path d="M0,0 C50,0 50,0 100,0 L100,62.932367149758456 C50,62.932367149758456 50,63 0,63 Z" fill="url(#gradient2.0.0)" fill-opacity="0.4"></path>
        
      
        <defs>
          
            <linearGradient id="gradient2.1.0">
              <stop offset="0%" stop-color="var(--primary-color)"></stop>
              <stop offset="100%" stop-color="red"></stop>
            </linearGradient>
          
        </defs>
        
          <path d="M0,68.5 C50,68.5 50,62.932367149758456 100,62.932367149758456 L100,63 C50,63 50,68.56766917293233 0,68.56766917293233 Z" fill="url(#gradient2.1.0)" fill-opacity="0.4"></path>
        
      
        <defs>
          
        </defs>
        
      
            </svg>
          </div>
      
          
          
          <div class="box type-entity" style="height:63px;">
            <div style="background-color:var(--primary-color);" title="2kW Varmtvann
Blaa" class="">
              
            </div>
            <div class="label" style="line-height:15px;">
                  
                  
                </div>
          </div>
          
        
          <div class="spacerv" style="height:5.5px;"></div>
          
          <div class="box type-entity" style="height:63px;">
            <div style="background-color:var(--primary-color);" title="2kÐ’ Avfukter" class="">
              
            </div>
            <div class="label" style="line-height:15px;">
                  
                  
                </div>
          </div>
          
        
          <div class="spacerv" style="height:5.5px;"></div>
          
          <div class="box type-entity" style="height:63px;">
            <div style="background-color:var(--primary-color);" title="2kW " class="">
              
            </div>
            <div class="label" style="line-height:15px;">
                  
                  
                </div>
          </div>
          
        
    </div>
  
    <div class="section" style="">
      <div class="connectors">
            <svg preserveAspectRatio="none" viewBox="0 0 100 200">
              
        <defs>
          
            <linearGradient id="gradient3.0.0">
              <stop offset="0%" stop-color="red"></stop>
              <stop offset="100%" stop-color="red"></stop>
            </linearGradient>
          
        </defs>
        
          <path d="M0,0 C50,0 50,0 100,0 L100,63 C50,63 50,63 0,63 Z" fill="url(#gradient3.0.0)" fill-opacity="0.4"></path>
        
      
            </svg>
          </div>
      
          
          
          <div class="box type-passthrough" style="height:63px;">
            <div style="background-color:red;" title="2kW " class="">
              
            </div>
            
          </div>
          
        
    </div>
  
    <div class="section" style="">
      
      
          
          
          <div class="box type-entity" style="height:63px;">
            <div style="background-color:red;" title="2kW " class="">
              
            </div>
            <div class="label" style="line-height:15px;">
                  
                  
                </div>
          </div>
          
        
    </div>
  
          </div>
        </ha-card>
      <style>
    ha-card {
        overflow-x: auto;
    }
    .container {
        display: flex;
        position: relative;
        /* height: 210px; */
        padding: 16px;
    }
    .container.with-header {
        padding-top: 0;
    }
    .section {
        flex: 1;
        flex-direction: column;
        position: relative;
        min-width: 0;
        max-width: 50%;
    }
    .wide .section:last-child {
        flex: initial;
    }
    .spacerv {
        transition: height 0.2s;
    }
    .box {
        display: flex;
        align-items: center;
        /* position: relative; */
        /* min-height: 1px; */
        transition: height 0.25s;
    }
    /* .box::before {
        content: "";
        position: absolute;
        top: -2px;
        bottom: -2px;
        left: -2px;
        right: -2px;
        background-color: var(--primary-color);
        opacity: 0.5;
        border-radius: 3px;
    } */
    .box div:first-child {
        display: flex;
        justify-content: center;
        align-items: center;
        overflow: hidden;
        background-color: var(--primary-color);
        width: 15px;
        height: 100%;
        cursor: pointer;
    }
    .box.type-passthrough div:first-child {
        opacity: 0.4;
    }
    .box.type-passthrough div.hl:first-child {
        opacity: 0.85;
    }
    .box .label {
        flex: 1;
        display: flex;
        align-items: center;
        padding: 0 10px;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }
    .box .label span {
        font-style: italic;
    }
    .connectors {
        position: absolute;
        top: 0;
        left: 15px;
        right: 0;
        height: 100%;
        overflow: hidden;
    }
    .connectors svg {
        position: absolute;
        left: -1px;
        width: 101%;
        height: 100%;
    }
</style>"
`;
